perception/autoware_tensorrt_bevdet/include/autoware/tensorrt_bevdet/bevdet.hpp

```diff
@@ -0,0 +1,240 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef AUTOWARE__TENSORRT_BEVDET__BEVDET_HPP_
+#define AUTOWARE__TENSORRT_BEVDET__BEVDET_HPP_
+
+#include "NvInfer.h"
+#include "common.hpp"
+#include "data.hpp"
+#include "postprocess.hpp"
+
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+
+#include <NvOnnxParser.h>
+#include <yaml-cpp/yaml.h>
+
+#include <algorithm>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace autoware::tensorrt_bevdet
+{
+class AdjFrame
+{
+public:
+  AdjFrame() {}
+  AdjFrame(int n, size_t buf_size)
+  : n_(n), buf_size_(buf_size), scenes_token_(n), ego2global_rot_(n), ego2global_trans_(n)
+  {
+    CHECK_CUDA(cudaMalloc(reinterpret_cast<void **>(&adj_buffer), n * buf_size));
+    CHECK_CUDA(cudaMemset(adj_buffer, 0, n * buf_size));
+    last_ = 0;
+    buffer_num_ = 0;
+    init_ = false;
+
+    for (auto & rot : ego2global_rot_) {
+      rot = Eigen::Quaternion<float>(0.f, 0.f, 0.f, 0.f);
+    }
+    for (auto & trans : ego2global_trans_) {
+      trans = Eigen::Translation3f(0.f, 0.f, 0.f);
+    }
+  }
+  const std::string & lastScenesToken() const { return scenes_token_[last_]; }
+
+  void reset()
+  {
+    last_ = 0;  // origin -1
+    buffer_num_ = 0;
+    init_ = false;
+  }
+
+  void saveFrameBuffer(
+    const void * curr_buffer, const std::string & curr_token,
+    const Eigen::Quaternion<float> & ego2global_rot, const Eigen::Translation3f & ego2global_trans)
+  {
+    int iters = init_ ? 1 : n_;
+    while (iters--) {
+      last_ = (last_ + 1) % n_;
+      CHECK_CUDA(cudaMemcpy(
+        reinterpret_cast<char *>(adj_buffer) + last_ * buf_size_, curr_buffer, buf_size_,
+        cudaMemcpyDeviceToDevice));
+      scenes_token_[last_] = curr_token;
+      ego2global_rot_[last_] = ego2global_rot;
+      ego2global_trans_[last_] = ego2global_trans;
+      buffer_num_ = std::min(buffer_num_ + 1, n_);
+    }
+    init_ = true;
+  }
+  int havingBuffer(int idx) { return static_cast<int>(idx < buffer_num_); }
+
+  const void * getFrameBuffer(int idx)
+  {
+    idx = (-idx + last_ + n_) % n_;
+    return reinterpret_cast<char *>(adj_buffer) + idx * buf_size_;
+  }
+  void getEgo2Global(
+    int idx, Eigen::Quaternion<float> & adj_ego2global_rot,
+    Eigen::Translation3f & adj_ego2global_trans)
+  {
+    idx = (-idx + last_ + n_) % n_;
+    adj_ego2global_rot = ego2global_rot_[idx];
+    adj_ego2global_trans = ego2global_trans_[idx];
+  }
+
```

Comment:
Consider adding const correctness to `getFrameBuffer` and `getEgo2Global` if the methods do not modify the internal state of the `AdjFrame` object.